### PR TITLE
SPALeRA Stochastic Gradient Descent

### PR DESCRIPTION
**Pull request by zoq**
Implementation of SPALeRA Stochastic Gradient Descent as proposed by A. Schoenauer-Sebag in "Stochastic Gradient Descent: Going As Fast As Possible But Not Faster".

### CODE REVIEW

File: src/mlpack/core/optimizers/spalera_sgd/spalera_stepsize.hpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+   */
+  bool Update(const double stepSize,
+              const double objective,
+              const size_t batchSize,
+              const size_t numFunctions,
+              arma::mat& iterate,
+              const arma::mat& gradient)
+  {
+    // The ratio of mini-batch size to training set size; needed for the
+    // Page-Hinkley relaxed objetive computations.
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/spalera_sgd/spalera_stepsize.hpp]: Small typo: `objective` :)

File: src/mlpack/core/optimizers/spalera_sgd/spalera_sgd_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+      }
+
+      // Ensure the last batch size isn't zero, to avoid division by zero before
+      // updating.
+      bool learningRateCheck = true;
+      if (lastBatchSize > 0)
+      {
+        // Now update the iterate.
+        learningRateCheck = updatePolicy.Update(stepSize, currentObjective,
+            lastBatchSize, numFunctions, iterate, gradient);
```
Comment (rcurtin) [inline src/mlpack/core/optimizers/spalera_sgd/spalera_sgd_impl.hpp]: If I am not mistaken, everything "special" about SPALeRA is contained in the `SPALeRAStepSize` class.  So could we refactor `SGD` to add a `DecayPolicy` type and then use `SGD<VanillaUpdate, SPALeRAStepSize>` as the SPALeRA optimizer?

**Comment by zoq** [comment]
> I guess the only question is how we handle this and the batch support PR (but I think that question exists for each of the four optimizer PRs you opened recently).

I'd like to merge this afterwards, I'll adjust the PR's once #1137 is merged, which I guess, should be soon.

File: src/mlpack/core/optimizers/spalera_sgd/spalera_sgd_impl.hpp (hunk 2 above)
Comment (zoq) [reply src/mlpack/core/optimizers/spalera_sgd/spalera_sgd_impl.hpp]: I thought about that, and discarded that idea, because the current policies just take ``iterate``, ``stepSize`` and ``gradient`` as parameter.
